libraries/tock-register-interface/src/peripheral.rs

```diff
@@ -0,0 +1,288 @@
+///         + tock_registers::Write<LongName = Trigger::Register>
+///     where
+///         Self: 's;
+///     fn aliased(&self) -> Self::aliased<'_>;
+/// }
+/// ```
+///
+/// # Padding
+/// If there is a gap between registers in a peripheral, you must declare that
+/// gap as a padding field:
+/// ```
+/// tock_registers::peripheral! {
+///     Foo {
+///         0x0 => register_a: u8 {},
+///
+///         // The padding field; specified by writing _ in place of the name
+///         // and omitting the data type and operation list.
+///         0x1 => _,
+///
+///         // The offset of the next field determines the size of the padding
+///         // (in this case, the padding is two bytes).
+///         0x3 => register_b: u8 {},
+///     }
+/// }
+/// ```
+/// Padding fields are ommitted from the generated trait.
+///
+/// # Doc comments
+/// ```
+/// tock_registers::peripheral! {
+///     /// You can put doc comments on peripheral declarations.
+///     Foo {
+///         /// You can also put doc comments on registers.
+///         0x0 => ctrl: u32 { Write },
+///     }
+/// }
+/// ```
+/// The doc comments will be emitted on the trait declaration, so they show up
+/// in rustdoc.
+///
+/// # Blanket Deref impl
+/// `tock_registers::peripheral!` emits a blanket implementation of the
+/// peripheral trait for types that `Deref` to an implementation of the
+/// peripheral:
+/// ```ignore
+/// impl<T: core::ops::Deref> Foo for T where T::Target: Foo {
+///     ...
+/// }
+/// ```
+/// Using this blanket impl, a driver that looks like:
+/// ```ignore
+/// struct FooDriver<F: Foo> {
+///     foo: F,
+/// }
+/// ```
+/// Can be used as either a `FooDriver<&FakeFoo>` or a `FooDriver<Rc<FakeFoo>>`
+/// in unit tests.
+///
+/// # Bus Adapters
+/// If you add the `#[allow_bus_adapter]` attribute to a peripheral:
+/// ```
+/// tock_registers::peripheral! {
+///     #[allow_bus_adapter]
+///     Foo {
+///         // ...
+///     }
+/// }
+/// ```
+/// the generated `RealFoo` structure will gain a
+/// `B: tock_registers::BusAdapter` generic argument:
+/// ```ignore
+/// #[derive(Clone, Copy)]
+/// struct RealFoo<
+///     P: tock_registers::MmioPointer,
+///     B: tock_registers::BusAdapter = tock_registers::DirectAdapter>
+///     { ... }
+/// ```
+/// This argument is used to support unusual chip designs (mainly some LiteX
+/// designs). Because the bus adapter can change the size of registers (e.g. a
+/// register with a data type of `u8` may take up more than 1 byte in the MMIO
+/// address space), `peripheral!` allows you to specify that an offset is to be
+/// inferred:
+/// ```
+/// tock_registers::peripheral! {
+///     #[allow_bus_adapter]
+///     Foo {
+///         // The first register is still at offset 0.
+///         0x0 => a: u8 { Read },
+///
+///         // This register has an inferred offset, which may depend on the bus
+///         // adapter passed to RealFoo.
+///         _ => b: u32 { Write },
```

Comment:
The more I think about this, the more complex it starts to feel to me. Does this mean that for something which is nominally
```
Foo {
   0x0 => zero: u8
   0x1 => one: u16
   0x3 => three: u8
   0x4 => four: u32
   0x8 => eight: u8
}
```
on a "byte-aligned bus" the struct would be unchanged, but otherwise
```
// halfword-aligned bus becomes

Foo {
   0x0 => zero: u8
   0x1 => _
   0x2 => one: u16
   0x4 => three: u8
   0x5 => _
   0x6 => four: u32
   0xa => eight: u8
}

// word-aligned bus becomes

Foo {
   0x0 => zero: u8
   0x1 => _
   0x2 => _
   0x3 => _
   0x4 => one: u16
   0x6 => _
   0x7 => _
   0x8 => three: u8
   0x9 => _
   0xa => _
   0xb => _
   0xc => four: u32
}
```

---

What happens if a struct like this has padding??
```
// byte-aligned bus
Nominal {
   0x0 => zero: u8
   0x1 => _: u8
   0x2 => two: u16
}

// half-word does... this?
ExpandExplicitPadding {
   0x0 => zero: u8
   0x1 => _  // from bus expansion
   0x2 => _: u8 // the originally defined padding
   0x3 => _ // from bus expansion
   0x4 => two: u16
}
// or this?
BusUsesExtantPadding {
   0x0 => zero: u8
   0x1 => _: u8 // originally defined padding meets bus expansion needs
   0x2 => two: u16
}
// but then a word-aligned bus would need to add padding no matter what?
```

---

I don't know what's better here, but the more I think about things, the more "inferred offset" scares me.

One, kind of verbose, concrete suggestion might be something like:
```
/// tock_registers::peripheral! {
///     #[allow_bus_adapter(u8,u32)]
///     Foo {
///         0x0,0x0 => a: u8 { Read },
///         0x1,0x4 => b: u8 { Read },
```
where the `allow_bus_adapter` explicitly states which alignments are allowed, and then the register placement definitions explicitly show how things are laid out, based on the underlying alignment?

Reply:
From tock-registers' perspective, the padding in LiteX registers is BusAdapters' problem, and does not show up in anything `peripheral!` does. `peripheral!` just sees that a register's size may differ from its data type's size.

My understanding is that LiteX registers do not have padding, @lschuermann to confirm.

To make your suggestion match my understanding, I *think* we could do:

```rust
use litex_registers::{C8B32, C32B32};

tock_registers::peripheral! {
    #[allow_bus_adapter(C8B32, C32B32)]
    Foo {
        [0x0, 0x0] => a: u16 { Read },
        [0x8, 0x4] => b: u8 { Read },
    }
}
```

But I would have to implement it to be sure.